pyquil/api/_qpu.py

```diff
@@ -154,3 +165,21 @@ def _build_patch_values(self) -> dict:
             patch_table[k.name][k.index] = v
 
         return patch_table
+
+    def _resolve_memory_references(self, expression: Expression) -> Union[float, int]:
+        """
+        :param expression:
+        """
+        if isinstance(expression, BinaryExp):
+            left = self._resolve_memory_references(expression.op1)
+            right = self._resolve_memory_references(expression.op2)
+            return expression.fn(left, right)
+        elif isinstance(expression, Function):
+            return expression.fn(_resolve_memory_references(expression.expression))
+        elif isinstance(expression, Parameter):
+            raise ValueError(f"Unexpected Parameter in gate expression: {expression}")
+        elif isinstance(expression, float) or isinstance(expression, int):
+            return expression
+        else:
+            assert isinstance(expression, MemoryReference)
```

Comment:
rather than asserting this, id rather you make this block an `elif` and use the `else` to throw an error. then, if you happened to omit some part of this "expression grammar" or it changes, you'll get a helpful error message instead of an `AssertionError`

Reply:
haha I also noticed that! thanks!